### PR TITLE
V11 release removal of node-fetch and undici dependencies

### PR DESCRIPTION
**Pull request by DellaBitta**
### Discussion

Our v11 release will require node 18+. Since fetch has been introduced in these node versions, we can remove our dependency on third party fetch implementations.

This PR removes our usage of `fetch` variants `undici` and `node-fetch` for our node target builds and our CI tools.

### Testing

CI.

### API Changes

NA.

### CODE REVIEW

File: packages/auth/src/platform_node/index.ts [hunk 1]
```diff
@@ -28,17 +28,12 @@ import { AuthImpl } from '../core/auth/auth_impl';
...
-  Response as undiciResponse
-} from 'undici';
 
 // Initialize the fetch polyfill, the types are slightly off so just cast and hope for the best
 FetchProvider.initialize(
-  undiciFetch as unknown as typeof fetch,
-  undiciHeaders as unknown as typeof Headers,
-  undiciResponse as unknown as typeof Response
+  fetch as unknown as typeof fetch,
+  Headers as unknown as typeof Headers,
```
Comment (hsubox76) [inline packages/auth/src/platform_node/index.ts:35]: Do we still have to do the casting or can it infer that it's a typeof, well, what it is?

File: packages/functions/src/index.node.ts [hunk 2]
```diff
@@ -15,9 +15,8 @@
  * limitations under the License.
  */
 import { registerFunctions } from './config';
-import { fetch as undiciFetch } from 'undici';
 
 export * from './api';
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-registerFunctions(undiciFetch as any, 'node');
+registerFunctions('node');
```
Comment (hsubox76) [inline packages/functions/src/index.node.ts:22]: Hm, I don't think we actually need a separate functions Node bundle anymore. Maybe that can be a separate PR though. The only thing different is that it labels the bundle "node" for platform logging, which is pointless if it's the same bundle. I guess it tells us how many node users are using functions. Also it doesn't export the types from public-types which is probably a mistake.

File: packages/auth-compat/index.node.ts [hunk 3]
```diff
@@ -23,15 +23,10 @@
...
 export * from './index';
 import { FetchProvider } from '@firebase/auth/internal';
-import {
-  fetch as undiciFetch,
-  Headers as undiciHeaders,
-  Response as undiciResponse
-} from 'undici';
 import './index';
 
 FetchProvider.initialize(
```
Comment (DellaBitta) [inline packages/auth-compat/index.node.ts:28]: I'm keeping the FetchProvider around because it's used in our testing framework.

File: packages/firestore/src/platform/browser_lite/fetch_connection.ts [hunk 4]
```diff
@@ -28,17 +27,6 @@ import { StringMap } from '../../util/types';
...
 export class FetchConnection extends RestConnection {
-  /**
-   * @param databaseInfo - The connection info.
-   * @param fetchImpl - `fetch` or a Polyfill that implements the fetch API.
-   */
-  constructor(
-    databaseInfo: DatabaseInfo,
-    private readonly fetchImpl: typeof fetch
-  ) {
-    super(databaseInfo);
```
Comment (DellaBitta) [inline packages/firestore/src/platform/browser_lite/fetch_connection.ts:39]: This consructor is removed because it would now have the same signature as RestConnection's constructor, so we just use that constructor instead.

File: packages/firestore/src/platform/node_lite/connection.ts [hunk 5]
```diff
@@ -25,8 +23,6 @@ export { newConnectivityMonitor } from '../browser/connection';
 
 /** Initializes the HTTP connection for the REST API. */
 export function newConnection(databaseInfo: DatabaseInfo): Connection {
-  // undici is meant to be API compatible with `fetch`, but its type doesn't
-  // match 100%.
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (DellaBitta) [inline packages/firestore/src/platform/node_lite/connection.ts:26]: I will remove this line.

File: packages/rules-unit-testing/test/util.test.ts [hunk 6]
```diff
@@ -165,7 +165,7 @@ describe('assertFails()', () => {
 
 describe('withFunctionTriggersDisabled()', () => {
   it('disabling function triggers does not throw, returns value', async function () {
-    const fetchSpy = sinon.spy(require('node-fetch'), 'default');
+    const fetchSpy = sinon.spy(globalThis, 'fetch');
```
Comment (hsubox76) [inline packages/rules-unit-testing/test/util.test.ts:168]: Does this work in Node 21.7.0+? https://github.com/firebase/firebase-js-sdk/pull/8492/files#diff-80382ef750022c8b0f79a66d4a5baf5fe0412100937528ef71e8938b62d312ffR22

Stub and mock won't, not 100% sure about spy but I think it also has to overwrite to spy. In any case, if there's a problem, I think what you gotta do is make an optional fetchImpl arg on withFunctionTriggersDisabled that defaults to `fetch` and then overwrite it here with a stub

File: packages/storage/test/node/connection.test.ts [hunk 7]
```diff
@@ -24,9 +24,9 @@ describe('Connections', () => {
   it('FetchConnection.send() should not reject on network errors', async () => {
     const connection = new FetchBytesConnection();
 
-    // need the casting here because fetch_ is a private member
-    stub(connection as any, 'fetch_').rejects();
+    const fetchStub = stub(globalThis, 'fetch').rejects();
```
Comment (hsubox76) [inline packages/storage/test/node/connection.test.ts:27]: I don't think this is gonna work in Node 21.7.0+ (see other comment).

File: scripts/emulator-testing/emulators/emulator.ts [hunk 8]
```diff
@@ -51,35 +50,59 @@ export abstract class Emulator {
...
+          (downloadComplete, downloadFailed) => {
+            fetch(this.binaryUrl)
+              .then(resp => {
+                if (resp.status !== 200) {
+                  console.log('Download of emulator failed: ', resp.statusText);
+                  downloadFailed();
                 }
-                resolve();
+                const reader = resp.body?.getReader();
+                reader?.read().then(function readChunk({ done, value }): any {
```
Comment (hsubox76) [inline scripts/emulator-testing/emulators/emulator.ts:70]: Is there a chance of this hanging if there's no reader or no body? Do we have to call a resolve() or reject() or downloadFailed() or something at the end as a failsafe?

**Review by hsubox76 — COMMENTED** [review]
Thanks, looks very thorough! Just a few questions.

File: packages/functions/src/index.node.ts (hunk 2 above)
Comment (DellaBitta) [reply packages/functions/src/index.node.ts:21]: I'll add the export.

File: packages/rules-unit-testing/test/util.test.ts (hunk 6 above)
Comment (hsubox76) [reply packages/rules-unit-testing/test/util.test.ts:168]: Looks like this is fine in Node now, resolving.

File: packages/functions/src/index.node.ts (hunk 2 above)
Comment (DellaBitta) [reply packages/functions/src/index.node.ts:22]: We spoke offline about removing the separate node bundle, will be done in a different PR.

File: scripts/emulator-testing/emulators/emulator.ts (hunk 8 above)
Comment (DellaBitta) [reply scripts/emulator-testing/emulators/emulator.ts]: Yes, thank you! Updated the code.

**Review by hsubox76 — APPROVED** [review]

File: packages/auth-compat/index.node.ts [hunk 9]
```diff
@@ -23,15 +23,10 @@
...
-  undiciFetch as unknown as typeof fetch,
-  undiciHeaders as unknown as typeof Headers,
-  undiciResponse as unknown as typeof Response
+  fetch as unknown as typeof fetch,
```
Comment (NhienLam) [inline packages/auth-compat/index.node.ts:29]: As Christina pointed out in the other comment, we also don't need casting here, correct?

**Review by NhienLam — CHANGES_REQUESTED** [review]
Auth changes LGTM

**Review by NhienLam — APPROVED** [review]

File: packages/auth-compat/index.node.ts (hunk 9 above)
Comment (DellaBitta) [reply packages/auth-compat/index.node.ts]: Done! Thanks!

**Review by MarkDuckworth — APPROVED** [review]
reviewed `packages/firestore/*`
